File: mobile/src/main/kotlin/de/scout/fireplace/network/NetworkModule.java
package de.scout.fireplace.network;

import dagger.Module;
import dagger.Provides;
import dagger.Reusable;
import de.scout.fireplace.BuildConfig;
import java.util.concurrent.TimeUnit;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;

import static de.scout.fireplace.network.NetworkClient.Type.ANONYMOUS;
import static de.scout.fireplace.network.NetworkClient.Type.AUTHENTICATED;

@Module
public abstract class NetworkModule {

  private static final int TIMEOUT_IN_SECONDS = 5;

  @Provides
  @NetworkAgent
  static String agent() {
    return "ANDROID";
  }

  @Provides
  static HttpLoggingInterceptor logging() {
    HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
    interceptor.setLevel(BuildConfig.DEBUG ? HttpLoggingInterceptor.Level.BODY : HttpLoggingInterceptor.Level.NONE);
    return interceptor;
  }

  @Provides
  @Reusable
  @NetworkClient(ANONYMOUS)
  static OkHttpClient anonymousClient(HeadersInterceptor headers, HttpLoggingInterceptor logging) {
    return new OkHttpClient.Builder()
        .connectTimeout(TIMEOUT_IN_SECONDS, TimeUnit.SECONDS)
        .readTimeout(TIMEOUT_IN_SECONDS, TimeUnit.SECONDS)
        .writeTimeout(TIMEOUT_IN_SECONDS, TimeUnit.SECONDS)
        .addInterceptor(headers)
        .addInterceptor(logging)
        .build();
  }

  @Provides
  @Reusable
  @NetworkClient(AUTHENTICATED)
  static OkHttpClient authenticatedClient(@NetworkClient(ANONYMOUS) OkHttpClient client, AuthenticationInterceptor interceptor) {
    return client.newBuilder()
        .addInterceptor(interceptor)
        .build();
  }
}

File: mobile/src/main/kotlin/de/scout/fireplace/network/NetworkModule.kt
package de.scout.fireplace.network

import dagger.Module
import dagger.Provides
import dagger.Reusable
import de.scout.fireplace.BuildConfig
import de.scout.fireplace.network.NetworkClient.Type.ANONYMOUS
import de.scout.fireplace.network.NetworkClient.Type.AUTHENTICATED
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import java.util.concurrent.TimeUnit

@Module
internal class NetworkModule {

  private val TIMEOUT_IN_SECONDS = 5

  @Provides
  @NetworkAgent
  fun agent() = "ANDROID"

  @Provides
  fun logging() = HttpLoggingInterceptor().apply { level = if (BuildConfig.DEBUG) HttpLoggingInterceptor.Level.BODY else HttpLoggingInterceptor.Level.NONE }

  @Provides
  @Reusable
  @NetworkClient(ANONYMOUS)
  fun anonymousClient(headers: HeadersInterceptor, logging: HttpLoggingInterceptor) = OkHttpClient.Builder()
      .connectTimeout(TIMEOUT_IN_SECONDS.toLong(), TimeUnit.SECONDS)
      .readTimeout(TIMEOUT_IN_SECONDS.toLong(), TimeUnit.SECONDS)
      .writeTimeout(TIMEOUT_IN_SECONDS.toLong(), TimeUnit.SECONDS)
      .addInterceptor(headers)
      .addInterceptor(logging)
      .build()

  @Provides
  @Reusable
  @NetworkClient(AUTHENTICATED)
  fun authenticatedClient(@NetworkClient(ANONYMOUS) client: OkHttpClient, interceptor: AuthenticationInterceptor) = client.newBuilder()
      .addInterceptor(interceptor)
      .build()
}
